src/dp_maglev.c

```diff
@@ -0,0 +1,210 @@
+// SPDX-FileCopyrightText: 2023 SAP SE or an SAP affiliate company and IronCore contributors
+// SPDX-License-Identifier: Apache-2.0
+
+#include <stdio.h>
+#include <stdlib.h>
+#include <stdbool.h>
+#include <string.h>
+#include <arpa/inet.h>
+#include "dp_error.h"
+#include "rte_rib6.h"
+#include "dp_maglev.h"
+
+#include <stdint.h>
+
+#define DP_MURMURHASH2_MAGIC	0x5bd1e995
+#define DP_DJB_HASH_MAGIC		5381
+#define DP_MAGLEV_POS_FREE		-1
+#define DP_SHIFT_UP				true
+#define DP_SHIFT_DOWN			false
+
+static uint32_t dp_murmur_hash2(const uint8_t ipv6[])
+{
+	uint32_t len = DP_IPV6_ADDR_SIZE;
+	const uint8_t *data = ipv6;
+	uint32_t h = 0 ^ len;
+	uint32_t k;
+
+	while (len >= 4) {
+		k = data[0] | (data[1] << 8) | (data[2] << 16) | (data[3] << 24);
+
+		k *= DP_MURMURHASH2_MAGIC;
+		k ^= k >> 24;
+		k *= DP_MURMURHASH2_MAGIC;
+
+		h *= DP_MURMURHASH2_MAGIC;
+		h ^= k;
+
+		data += 4;
+		len -= 4;
+	}
+
+	switch (len) {
+	case 3:
+		h ^= data[2] << 16; // Fall through
+	case 2:
+		h ^= data[1] << 8;  // Fall through
+	case 1:
+		h ^= data[0];
+		h *= DP_MURMURHASH2_MAGIC;
+		break;
+	default:
+		break;
+	}
+
+	h ^= h >> 13;
+	h *= DP_MURMURHASH2_MAGIC;
+	h ^= h >> 15;
+
+	return h;
+}
+
+static uint32_t dp_djb_hash(const uint8_t ipv6[])
+{
+	uint32_t hash = DP_DJB_HASH_MAGIC;
+	const uint8_t *byte = ipv6;
+	int c;
+
+	while ((c = *byte++))
+		hash = ((hash << 5) + hash) + c;
+
+	hash &= ~(1 << 31);
+	return hash;
+}
+
+static int *dp_maglev_permutation(struct lb_value *lbval)
+{
+	int  i, j, size, *permutation;
+	uint32_t offset, skip;
+
+	size = lbval->back_end_cnt * DP_LB_MAGLEV_LOOKUP_SIZE;
+	permutation = rte_zmalloc("maglev_perm", sizeof(int) * size, RTE_CACHE_LINE_SIZE);
+
+	if (!permutation)
+		return NULL;
+
+	for (i = 0; i < lbval->back_end_cnt; i++) {
+		offset = dp_murmur_hash2(lbval->back_end_ips[i]) % DP_LB_MAGLEV_LOOKUP_SIZE;
+		skip = (dp_djb_hash(lbval->back_end_ips[i]) % (DP_LB_MAGLEV_LOOKUP_SIZE - 1)) + 1;
+
+		for (j = 0; j < DP_LB_MAGLEV_LOOKUP_SIZE; j++)
+			permutation[i * DP_LB_MAGLEV_LOOKUP_SIZE + j] = (offset + j * skip) % DP_LB_MAGLEV_LOOKUP_SIZE;
+	}
+	return permutation;
+}
+
+static int dp_maglev_populate(struct lb_value *lbval, int *permutation)
+{
+	int *next = rte_zmalloc("maglev_hash", lbval->back_end_cnt * sizeof(int), RTE_CACHE_LINE_SIZE);
+	int i, j, pos, num = 0;
+
+	if (!next)
+		return DP_ERROR;
+
+	for (i = 0; i < DP_LB_MAGLEV_LOOKUP_SIZE; i++)
+		lbval->maglev_hash[i] = DP_MAGLEV_POS_FREE;
+
+	while (true) {
+		for (i = 0; i < lbval->back_end_cnt; i++) {
+			for (j = next[i]; j < DP_LB_MAGLEV_LOOKUP_SIZE; j++) {
+				pos = permutation[i*DP_LB_MAGLEV_LOOKUP_SIZE + j];
+				next[i]++;
+				if (lbval->maglev_hash[pos] == DP_MAGLEV_POS_FREE) {
+					lbval->maglev_hash[pos] = (int16_t)i;
+					num++;
+					if (num == DP_LB_MAGLEV_LOOKUP_SIZE)
+						goto out;
+					break;
+				}
+			}
+		}
+	}
```

Comment:
I guess this is Google's algorithm, but still, an infinite loop just raises alarms for me.

Maybe having an always inline `dp_maglev_populate_iteration()` and then `while (dp_maglev_populate_iteration())` which also eliminates the `goto` would make this seem a bit more tame?

But if this is more of a 1:1 code from Google and there is a reference to such a document, then it would be best left as-is to resemble the original.

Reply:
Done